sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -295,12 +299,12 @@ class Dataset[T] private[sql](
    */
   def groupBy[K : Encoder](func: T => K): GroupedDataset[K, T] = {
     val inputPlan = queryExecution.analyzed
-    val withGroupingKey = AppendColumn(func, inputPlan)
+    val withGroupingKey = AppendColumn(func, resolvedTEncoder, inputPlan)
```

Comment:
My thoughts are that we should follow the same pattern as in expressions since we are already have to do eager resolution here inorder to output errors.  Logical query plans after the analyzer have resolved encoders, and then let the physical query plans do the binding.  The nice thing about passing a resolved encoder in is that in some cases the resolution could be ambiguous (think when we have both grouping columns and data columns with overlapping field names) even though I don't think that is the case here.

That said, we should maybe consider moving this step into the analyzer at some point.